R/tools.R

```diff
@@ -26,28 +26,49 @@ lint_dir <- function(path) {
   lints
 }
 
+lint_path <- function(path) {
+  if (fs::is_dir(path)) {
+    lint_dir(path)
+  } else {
+    if (interactive()) {
+      message(cli::format_inline("Linting {.file {path}}"))
+    }
+    lintr::lint(filename = path)
+  }
+}
+
 #' Lint R
 #'
 #' Uses the `{lintr}` package to check all R sources in the `app` and `tests/testthat` directories
 #' for style errors.
 #'
 #' The linter rules can be adjusted in the `.lintr` file.
 #'
-#' You can set the maximum number of accepted style errors
-#' with the `legacy_max_lint_r_errors` option in `rhino.yml`.
-#' This can be useful when inheriting legacy code with multiple styling issues.
+#' You can set the maximum number of accepted style errors with the `legacy_max_lint_r_errors`
+#' option in `rhino.yml`. This can be useful when inheriting legacy code with multiple styling
+#' issues.
+#'
+#' @param paths a vector paths to directories or `.R` files; defaults to `NULL` and checks `app` and
+#'   `tests/testthat` directories
 #'
 #' @return None. This function is called for side effects.
 #'
 #' @export
-lint_r <- function() {
-  max_errors <- read_config()$legacy_max_lint_r_errors
+lint_r <- function(paths = NULL) {
+  max_errors <- max_errors <- read_config()$legacy_max_lint_r_errors
   if (is.null(max_errors)) max_errors <- 0
 
-  lints <- c(
-    lint_dir("app"),
-    lint_dir(fs::path("tests", "testthat"))
+  if (is.null(paths)) {
+    paths <- c("app", "tests/testthat")
+  }
+
+  lints <- purrr::map(
+    paths,
+    ~ c(lint_path(.x))
   )
+
+  lints <- purrr::reduce(lints, c)
+
```

Comment:
How about this less verbose version with vanilla Shiny? :wink:
```suggestion
  lints <- do.call(c, lapply(paths, lint_path))
```

It should also perform better: with `do.call` we only make one call to `c`, while with `purrr:reduce` we essentially append elements to vector one by one (could become noticeable with thousands of linter errors, as sometimes happens when migrating legacy code).

Reply:
This does look better. I was already thinking of using `lapply()` instead of `purrr::map()`, but I forgot about `do.call()`.